File: openmrs-client/src/main/java/org/openmrs/mobile/api/repository/ConceptRepository.java
package org.openmrs.mobile.api.repository;

import org.openmrs.mobile.api.RestApi;
import org.openmrs.mobile.api.RestServiceBuilder;
import org.openmrs.mobile.dao.ConceptDAO;
import org.openmrs.mobile.models.Concept;
import org.openmrs.mobile.models.Results;
import org.openmrs.mobile.utilities.ApplicationConstants;

import java.util.List;

import retrofit2.Call;

public class ConceptRepository extends RetrofitRepository {

    private final RestApi api;
    private final ConceptDAO cache;

    public ConceptRepository() {
        this.api = RestServiceBuilder.createService(RestApi.class);
        this.cache = new ConceptDAO();
    }

    public Call<Results<Concept>> getConcepts(String representation, String searchQuery, Integer limit, Integer startIndex) {
        return api.getConcepts(representation, searchQuery, limit, startIndex);
    }

    public List<Concept> getCachedConcepts(String searchQuery) {
        return cache.findConceptsByName(searchQuery);
    }

    public Call<Results<Concept>> searchConcepts(String query) {
        return api.getConceptRefTerm(ApplicationConstants.API.FULL, query);
    }
}

File: openmrs-android-sdk/src/main/java/com/openmrs/android_sdk/library/api/repository/ConceptRepository.kt
package com.openmrs.android_sdk.library.api.repository

import com.openmrs.android_sdk.library.dao.ConceptRoomDAO
import com.openmrs.android_sdk.library.databases.AppDatabaseHelper
import com.openmrs.android_sdk.library.models.ConceptAnswers
import com.openmrs.android_sdk.library.models.ConceptMembers
import com.openmrs.android_sdk.library.models.SystemProperty
import com.openmrs.android_sdk.utilities.ApplicationConstants.API.FULL
import rx.Observable
import javax.inject.Inject
import javax.inject.Singleton
import java.util.concurrent.Callable

@Singleton
class ConceptRepository @Inject constructor(private val conceptRoomDAO: ConceptRoomDAO) : BaseRepository() {

    /**
     * Gets system property.
     *
     * @param systemProperty the system property
     * @return the system property
     */
    fun getSystemProperty(systemProperty: String): Observable<SystemProperty> {
        return AppDatabaseHelper.createObservableIO(Callable {
            restApi.getSystemProperty(systemProperty, FULL).execute().run {
                if (isSuccessful && body() != null) return@Callable body()!!.results[0]
                else throw Exception("Error fetching concepts: ${message()}")
            }
        })
    }

    /**
     * Get concept answers by UUID
     *
     * @param uuid UUID of the concept
     * @return Observable ConceptAnswers
     */
    fun getConceptByUuid(uuid: String): Observable<ConceptAnswers> {
        return AppDatabaseHelper.createObservableIO(Callable {
            restApi.getConceptFromUUID(uuid).execute().run {
                if (isSuccessful && body() != null) return@Callable body()!!
                else throw Exception("Error fetching concepts by uuid: ${message()}")
            }
        })
    }

    /**
     * Gets concept members.
     *
     * @param uuid the uuid of the concept
     * @return the concept members
     */
    fun getConceptMembers(uuid: String): Observable<ConceptMembers> {
        return AppDatabaseHelper.createObservableIO(Callable {
            restApi.getConceptMembersFromUUID(uuid).execute().run {
                if (isSuccessful && body() != null) return@Callable body()!!
                else throw Exception("Error fetching concept members: ${message()}")
            }
        })
    }


    fun getConceptCountFromDb() = conceptRoomDAO.conceptsCount

}
